File: ios/FiHavenCore/Sources/FiHavenCore/Logic/SubscriptionsFinder.swift
import Foundation

/// Subscription finder + manage links — port of subscriptionsFinder.js.
public enum SubscriptionsFinder {
    public static let staleDays = 60
    public static let trialReminderDays = 3

    public struct Item: Identifiable, Equatable, Sendable {
        public var id: String
        public var billId: String?
        public var merchantKey: String
        public var name: String
        public var monthly: Double
        public var amount: Double
        public var source: String
        public var lastDate: String?
        public var priceUp: Double?
        public var stale: Bool
        public var nextDue: Date?
        public var manageUrl: String?
        public var trialEnds: String?
        public var trialDaysLeft: Int?
        public var trialSoon: Bool
        public var duplicate: Bool
    }

    public static let amountSimilarity = 0.15

    public static func monthlyOfBill(_ b: Bill) -> Double {
        switch b.frequency {
        case "Weekly": return b.amount * 52 / 12
        case "Bi-weekly": return b.amount * 26 / 12
        case "Quarterly": return b.amount / 3
        case "Annually": return b.amount / 12
        default: return b.amount
        }
    }

    public static func amountsSimilar(_ amts: [Double]) -> Bool {
        guard amts.count >= 2 else { return true }
        let minA = amts.map { abs($0) }.min() ?? 0
        let maxA = amts.map { abs($0) }.max() ?? 0
        if maxA < 0.005 { return true }
        return (maxA - minA) / maxA <= amountSimilarity
    }

    public static func build(
        bills: [Bill],
        transactions: [SpendTransaction],
        tz: TimeZone,
        declined: [String] = []
    ) -> [Item] {
        let declinedSet = Set(declined.map { $0.lowercased() }.filter { !$0.isEmpty })
        var trackedKeys = Set<String>()
        var out: [Item] = []
        for b in bills where b.category == "Subscriptions" && !b.archived && !DateLogic.billEnded(b, tz: tz) {
            let left = trialDaysLeft(b.trialEnds, tz: tz)
            let name = b.name.isEmpty ? "Subscription" : b.name
            let mk = SubscriptionLinks.normalizeKey(name)
            if !mk.isEmpty { trackedKeys.insert(mk) }
            out.append(Item(
                id: "bill-\(b.id)",
                billId: b.id,
                merchantKey: mk,
                name: name,
                monthly: monthlyOfBill(b),
                amount: b.amount,
                source: "bill",
                lastDate: nil,
                priceUp: nil,
                stale: false,
                nextDue: BillSchedule.nextDueDate(b, tz: tz),
                manageUrl: SubscriptionLinks.manageUrl(for: b),
                trialEnds: b.trialEnds,
                trialDaysLeft: left,
                trialSoon: left.map { $0 >= 0 && $0 <= trialReminderDays } ?? false,
                duplicate: false
            ))
        }
        let withMerchant = transactions.filter { !$0.merchant.trimmingCharacters(in: .whitespaces).isEmpty }
        let byMerchant = Dictionary(grouping: withMerchant) { $0.merchant.trimmingCharacters(in: .whitespaces).lowercased() }
        for (_, list) in byMerchant {
            let months = Set(list.map { String($0.date.prefix(7)) })
            if months.count < 2 || list.count < 2 { continue }
            if months.count < 3 && !amountsSimilar(list.map(\.amount)) { continue }
            let sorted = list.sorted { $0.date < $1.date }
            guard let latest = sorted.last else { continue }
            let minAmt = list.map(\.amount).min() ?? 0
            let days = daysSince(latest.date, tz: tz) ?? 0
            let mk = SubscriptionLinks.normalizeKey(latest.merchant)
            if !mk.isEmpty && (declinedSet.contains(mk) || trackedKeys.contains(mk)) { continue }
            out.append(Item(
                id: "tx-\(latest.merchant)",
                billId: nil,
                merchantKey: mk,
                name: latest.merchant,
                monthly: latest.amount,
                amount: latest.amount,
                source: "tx",
                lastDate: latest.date,
                priceUp: latest.amount > minAmt + 0.005 ? minAmt : nil,
                stale: days > staleDays,
                nextDue: nil,
                manageUrl: SubscriptionLinks.manageUrl(for: Bill(id: "", name: latest.merchant, business: latest.merchant)),
                trialEnds: nil,
                trialDaysLeft: nil,
                trialSoon: false,
                duplicate: false
            ))
        }
        let dupes = duplicateKeys(out)
        return out.map { var i = $0; i.duplicate = dupes.contains(i.id); return i }
            .sorted { $0.monthly > $1.monthly }
    }

    private static func duplicateKeys(_ items: [Item]) -> Set<String> {
        var byKey: [String: [Item]] = [:]
        items.forEach { i in
            let k = SubscriptionLinks.normalizeKey(i.name)
            guard !k.isEmpty else { return }
            byKey[k, default: []].append(i)
        }
        return Set(byKey.filter { $0.value.count > 1 }.flatMap { $0.value.map(\.id) })
    }

    private static func trialDaysLeft(_ trialEnds: String?, tz: TimeZone) -> Int? {
        guard let trialEnds, trialEnds.range(of: #"^\d{4}-\d{2}-\d{2}$"#, options: .regularExpression) != nil,
              let end = DateLogic.parseDate(trialEnds, tz: tz) else { return nil }
        let today = DateLogic.today(tz: tz)
        return Calendar.current.dateComponents([.day], from: today, to: end).day
    }

    private static func daysSince(_ iso: String, tz: TimeZone) -> Int? {
        guard let d = DateLogic.parseDate(iso, tz: tz) else { return nil }
        return Calendar.current.dateComponents([.day], from: d, to: Date()).day
    }
}

public enum SubscriptionLinks {
    private static let urls: [String: String] = [
        "netflix": "https://www.netflix.com/cancelplan",
        "spotify": "https://www.spotify.com/account/subscription/",
        "hulu": "https://secure.hulu.com/account",
        "disneyplus": "https://www.disneyplus.com/account",
        "max": "https://www.max.com/account",
        "amazon": "https://www.amazon.com/gp/mprimecentral",
        "prime": "https://www.amazon.com/gp/mprimecentral",
        "youtube": "https://www.youtube.com/paid_memberships",
        "adobe": "https://account.adobe.com/plans",
    ]

    public static func normalizeKey(_ name: String) -> String {
        name.lowercased().replacingOccurrences(of: "[^a-z0-9]+", with: "", options: .regularExpression)
    }

    public static func manageUrl(for bill: Bill) -> String? {
        if let u = extractUrl(bill.notes) { return u }
        for raw in [bill.business, bill.name].compactMap({ $0?.trimmingCharacters(in: .whitespaces) }).filter({ !$0.isEmpty }) {
            let key = normalizeKey(raw)
            if let direct = urls[key] { return direct }
            if let hit = urls.keys.first(where: { key.contains($0) || $0.contains(key) }) { return urls[hit] }
        }
        return nil
    }

    private static func extractUrl(_ notes: String) -> String? {
        guard let m = notes.range(of: #"https?://[^\s]+"#, options: .regularExpression) else { return nil }
        return String(notes[m]).trimmingCharacters(in: CharacterSet(charactersIn: ",.)"))
    }
}

File: android/core/src/main/kotlin/app/fihaven/core/logic/SubscriptionsFinder.kt
package app.fihaven.core.logic

import app.fihaven.core.model.Bill
import app.fihaven.core.model.SpendTransaction
import java.time.LocalDate
import java.time.ZoneId
import java.time.temporal.ChronoUnit

/// Subscription finder logic — port of subscriptionsFinder.js + subscriptionLinks.js.
object SubscriptionsFinder {
    const val STALE_DAYS = 60L
    const val TRIAL_REMINDER_DAYS = 3L
    const val AMOUNT_SIMILARITY = 0.15

    data class Item(
        val id: String,
        val billId: String?,
        val merchantKey: String,
        val name: String,
        val monthly: Double,
        val amount: Double,
        val source: String,
        val lastDate: String?,
        val priceUp: Double?,
        val stale: Boolean,
        val nextDue: LocalDate?,
        val manageUrl: String?,
        val trialEnds: String?,
        val trialDaysLeft: Long?,
        val trialSoon: Boolean,
        val duplicate: Boolean,
    )

    fun monthlyOfBill(b: Bill): Double = when (b.frequency) {
        "Weekly" -> b.amount * 52 / 12
        "Bi-weekly" -> b.amount * 26 / 12
        "Quarterly" -> b.amount / 3
        "Annually" -> b.amount / 12
        else -> b.amount
    }

    /** True when amounts are within [AMOUNT_SIMILARITY] of each other (relative to max). */
    fun amountsSimilar(amts: List<Double>): Boolean {
        if (amts.size < 2) return true
        val min = amts.minOf { kotlin.math.abs(it) }
        val max = amts.maxOf { kotlin.math.abs(it) }
        if (max < 0.005) return true
        return (max - min) / max <= AMOUNT_SIMILARITY
    }

    fun build(
        bills: List<Bill>,
        transactions: List<SpendTransaction>,
        zone: ZoneId,
        declined: List<String> = emptyList(),
    ): List<Item> {
        val declinedSet = declined.map { it.lowercase() }.filter { it.isNotBlank() }.toSet()
        val trackedKeys = mutableSetOf<String>()
        val out = mutableListOf<Item>()

        bills.filter { it.category == "Subscriptions" && !it.archived && !DateLogic.billEnded(it, zone) }
            .forEach { b ->
                val trial = b.trialEnds
                val left = trialDaysLeft(trial, zone)
                val name = b.name.ifBlank { "Subscription" }
                val mk = SubscriptionLinks.normalizeKey(name)
                if (mk.isNotBlank()) trackedKeys += mk
                out += Item(
                    id = "bill-${b.id}",
                    billId = b.id,
                    merchantKey = mk,
                    name = name,
                    monthly = monthlyOfBill(b),
                    amount = b.amount,
                    source = "bill",
                    lastDate = null,
                    priceUp = null,
                    stale = false,
                    nextDue = BillSchedule.nextDueDate(b, zone),
                    manageUrl = SubscriptionLinks.manageUrl(b),
                    trialEnds = trial,
                    trialDaysLeft = left,
                    trialSoon = left != null && left in 0..TRIAL_REMINDER_DAYS,
                    duplicate = false,
                )
            }

        transactions.filter { it.merchant.trim().isNotEmpty() }
            .groupBy { it.merchant.trim().lowercase() }
            .forEach { (_, list) ->
                val months = list.map { it.date.take(7) }.toSet()
                if (months.size < 2 || list.size < 2) return@forEach
                if (months.size < 3 && !amountsSimilar(list.map { it.amount })) return@forEach

                val sorted = list.sortedBy { it.date }
                val latest = sorted.last()
                val minAmt = list.minOf { it.amount }
                val days = DateLogic.parseDate(latest.date)?.let {
                    ChronoUnit.DAYS.between(it, DateLogic.today(zone))
                } ?: 0L
                val mk = SubscriptionLinks.normalizeKey(latest.merchant)
                if (mk.isNotBlank() && (declinedSet.contains(mk) || trackedKeys.contains(mk))) return@forEach

                out += Item(
                    id = "tx-${latest.merchant}",
                    billId = null,
                    merchantKey = mk,
                    name = latest.merchant,
                    monthly = latest.amount,
                    amount = latest.amount,
                    source = "tx",
                    lastDate = latest.date,
                    priceUp = if (latest.amount > minAmt + 0.005) minAmt else null,
                    stale = days > STALE_DAYS,
                    nextDue = null,
                    manageUrl = SubscriptionLinks.manageUrl(Bill(name = latest.merchant, business = latest.merchant)),
                    trialEnds = null,
                    trialDaysLeft = null,
                    trialSoon = false,
                    duplicate = false,
                )
            }

        val dupes = duplicateKeys(out)
        return out.map { it.copy(duplicate = dupes.contains(it.id)) }.sortedByDescending { it.monthly }
    }

    private fun duplicateKeys(items: List<Item>): Set<String> {
        val byKey = items.groupBy { SubscriptionLinks.normalizeKey(it.name) }
        return byKey.filter { it.key.isNotBlank() && it.value.size > 1 }
            .flatMap { it.value.map { i -> i.id } }
            .toSet()
    }

    private fun trialDaysLeft(trialEnds: String?, zone: ZoneId): Long? {
        if (trialEnds.isNullOrBlank() || !trialEnds.matches(Regex("""^\d{4}-\d{2}-\d{2}$"""))) return null
        val end = DateLogic.parseDate(trialEnds) ?: return null
        val today = DateLogic.today(zone)
        return ChronoUnit.DAYS.between(today, end)
    }
}

object SubscriptionLinks {
    private val urls = mapOf(
        "netflix" to "https://www.netflix.com/cancelplan",
        "spotify" to "https://www.spotify.com/account/subscription/",
        "hulu" to "https://secure.hulu.com/account",
        "disneyplus" to "https://www.disneyplus.com/account",
        "max" to "https://www.max.com/account",
        "amazon" to "https://www.amazon.com/gp/mprimecentral",
        "prime" to "https://www.amazon.com/gp/mprimecentral",
        "youtube" to "https://www.youtube.com/paid_memberships",
        "adobe" to "https://account.adobe.com/plans",
    )

    fun normalizeKey(name: String): String =
        name.lowercase().replace(Regex("[^a-z0-9]+"), "")

    fun manageUrl(bill: Bill): String? {
        extractUrl(bill.notes)?.let { return it }
        listOf(bill.business, bill.name).mapNotNull { it?.trim() }.filter { it.isNotEmpty() }.forEach { n ->
            val key = normalizeKey(n)
            urls[key]?.let { return it }
            urls.keys.firstOrNull { key.contains(it) || it.contains(key) }?.let { return urls[it] }
        }
        return null
    }

    private fun extractUrl(notes: String): String? {
        val m = Regex("""https?://[^\s]+""").find(notes) ?: return null
        return m.value.trimEnd(',', '.', ')')
    }
}
